lib/matplotlib/tests/test_collections.py

```diff
@@ -649,6 +649,19 @@ def test_collection_set_verts_array():
         assert np.array_equal(ap._codes, atp._codes)
 
 
+def test_collection_set_array():
+    xs = np.arange(10)
+    ymin = 0
+    ymax = 1
+    vals = [*range(10)]
+    # Test set_array with list
+    plot = plt.vlines(xs, ymin, ymax, array=vals)
```

Comment:
This tests the line problem. But can you test `set_array` directly?  I'm still unclear when you would use it outside this indirect use.  If its only goal is to be used indirectly maybe it should just be private?

Reply:
Testing `set_array` something like this?
```python
    xs = np.arange(10)
    ymin = 0
    ymax = 1
    vals = [*range(10)]
    plot = plt.vlines(xs, ymin, ymax)
    
    # Test set_array with list
    plot.set_array(vals)

    # Test if array kwarg is copied
    vals[5] = 45
    assert np.not_equal(vals, plot.get_array()).any()
    
```

Reply:
I don't know ;-)  I have no idea what set_array is meant to do in the context of a scalar mappable, so thats what I was hoping could be added.  For instance the test above is great, but I am still not sure what this will do to the plot.  I could download the PR and see, but...

Reply:
I'm not sure if we use ScalarMappable class directly anywhere, although classes which inherit from it do use `set_array` directly, for example:
https://github.com/matplotlib/matplotlib/blob/67aa50d164669a1340fe05e06dfc7223e25c87b3/examples/shapes_and_collections/line_collection.py#L76-L79
https://github.com/matplotlib/matplotlib/blob/67aa50d164669a1340fe05e06dfc7223e25c87b3/examples/shapes_and_collections/artist_reference.py#L94-L95

Reply:
> I have no idea what set_array is meant to do in the context of a scalar mappable

ScalarMappable basically would map scalar data, in this case `vals`, into a color map format (RGBA). The result would induce colors to the artists w.r.t the scalars provided.
So in case of `vlines` (I realise now that `test_collection_set_array` should not have reference to `vlines`, and instead test `Collection` class' inherited `set_array` function directly), the resultant plot will have lines colored (the referenced issue https://github.com/matplotlib/matplotlib/issues/18841#issue-732646421 has that plot) according to the array `vals`.

Reply:
I updated the PR summary, hopefully making this a bit easier to understand.